modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx

```diff
@@ -8,48 +8,160 @@ import {TreeView} from '@clayui/core';
 import {ClayCheckbox} from '@clayui/form';
 import ClayLoadingIndicator from '@clayui/loading-indicator';
 import ClayModal from '@clayui/modal';
+import {fetch} from 'frontend-js-web';
 import React, {useEffect, useState} from 'react';
 
 import {FDSViewType} from '../../../FDSViews';
 import {getFields} from '../../../api';
 import {IField} from '../../../types';
+import openDefaultFailureToast from '../../../utils/openDefaultFailureToast';
+import openDefaultSuccessToast from '../../../utils/openDefaultSuccessToast';
 import {IFDSField} from '../Fields';
 
+interface IFieldTreeItem extends IField {
+	children?: IFieldTreeItem[];
+	savedId?: number;
+	selected?: boolean;
+}
+
+function visit(fields: Array<IFieldTreeItem>, callback: Function) {
+	fields.forEach((field) => {
+		callback(field);
+
+		if (field.children) {
+			visit(field.children, callback);
+		}
+	});
+}
+
+const applySavedFDSFields = ({
+	fields: initialFields,
+	savedFDSFields,
+}: {
+	fields: IField[];
+	savedFDSFields: Array<IFDSField>;
+}): [Set<React.Key>, Array<IFieldTreeItem>] => {
+	const selectedKeys = new Set<React.Key>();
+	const fields: IFieldTreeItem[] = Array.from(initialFields);
+
+	visit(fields, (field: IFieldTreeItem) => {
+		const savedFDSField = savedFDSFields.find(
+			(savedFDSField) => savedFDSField.name === field.name
+		);
+
+		if (savedFDSField) {
+			selectedKeys.add(savedFDSField.name);
+
+			field.savedId = savedFDSField.id;
+		}
+
+		field.id = field.name;
+	});
+
+	return [selectedKeys, fields];
+};
```

Comment:
There was a problem here with the `id` property. It is used by `ClayTreeView` as a unique identifier, in its internal selection mechanism. In our case, ID is set only if the field is saved to DSM persistence. If it's a new field taken from OpenAPI, it does not have an ID. That's why it cannot be used as a unique identifier. Instead, we are setting `ID` property to have `name` value, as `name` is unique. This is local change, caused by and specific only to the `ClayTreeView` component.

Reply:
It sounds strange having a couple of similar lists with different types (`deletionIds`) :shrug: [here](https://github.com/liferay-frontend/liferay-portal/pull/3806/files#diff-3bdfde06fddcbc2f5c626b00d67c7727b00ae4df1d5ca558a02f71edd78aaf50R162) and [here](https://github.com/liferay-frontend/liferay-portal/pull/3806/files#diff-3bdfde06fddcbc2f5c626b00d67c7727b00ae4df1d5ca558a02f71edd78aaf50R162) but as you mention, the scope is limited.

Reply:
@juanjofgliferay Your intuition was right on this one! The inconsistency resulted in the bug from https://github.com/liferay-frontend/liferay-portal/pull/3806#issuecomment-1808246312, where we ended up comparing strings and integers. My bad for not looking into this more thoroughly. Also, a lesson for me: don't be lazy with setting `Function` as TS type, argument and return types would have prevented this.